src/mlpack/methods/ne/neat.hpp

```diff
@@ -0,0 +1,974 @@
+        if (idxInChild == -1) {
+          childGenome.AddHiddenNeuron(dadGenome.aNeuronGenes[idxInParent]);
+        }
+////printf("crossover 11\n");
+        // Add to neuron
+        idxInChild = childGenome.GetNeuronIndex(dadGenome.aLinkGenes[idx].ToNeuronId());
+        idxInParent = dadGenome.GetNeuronIndex(dadGenome.aLinkGenes[idx].ToNeuronId());
+        if (idxInChild == -1) {
+          childGenome.AddHiddenNeuron(dadGenome.aNeuronGenes[idxInParent]);
+        }
+////printf("crossover 12\n");
+        continue;
+      }  
+    }
+  }
+
+  // Crossover two genome to get one genome.
+  void Crossover(Genome& genome1, Genome& genome2, Genome& childGenome) {
+    if (CompareGenome(genome1, genome2)) {  // genome1 is better
+      CrossoverLinkAndNeuron(genome1, genome2, childGenome);
+    } else {
+      CrossoverLinkAndNeuron(genome2, genome1, childGenome);
+    }
+  }
+
+  // Measure two genomes' disjoint (including exceed).
+  // NOTICE: we can separate into disjoint and exceed. But currently maybe it is enough.
+  double Disjoint(Genome& genome1, Genome& genome2) {
+    double numDisjoint = 0;
+
+    for (ssize_t i=0; i<genome1.NumLink(); ++i) {
+      ssize_t innovId = genome1.aLinkGenes[i].InnovationId();
+      bool linkContainedInGenome2 = genome2.ContainLink(innovId);
+      if (!linkContainedInGenome2) {
+        ++numDisjoint;
+      } 
+    }
+
+    for (ssize_t i=0; i<genome2.NumLink(); ++i) {
+      ssize_t innovId = genome2.aLinkGenes[i].InnovationId();
+      bool linkContainedInGenome1 = genome1.ContainLink(innovId);
+      if (!linkContainedInGenome1) {
+        ++numDisjoint;
+      }
+    }
+
+    ssize_t largerGenomeSize = std::max(genome1.NumLink(), genome2.NumLink());
+    double deltaD = numDisjoint / largerGenomeSize;
+    return deltaD; 
+  }
+
+  // Measure two genomes' weight difference.
+  double WeightDiff(Genome& genome1, Genome& genome2) {
+    double deltaW = 0;
+    ssize_t coincident = 0;
+
+    for (ssize_t i=0; i<genome1.NumLink(); ++i) {
+      int linkEnabledInGenome1 = (int) genome1.aLinkGenes[i].Enabled();
+      ssize_t innovId = genome1.aLinkGenes[i].InnovationId();
+      ssize_t idx = genome2.GetLinkIndex(innovId);
+      bool linkContainedInGenome2 = (idx != -1);
+
+      if (linkContainedInGenome2) {
+        int linkEnabledInGenome2 = (int) genome2.aLinkGenes[idx].Enabled();
+        deltaW += std::abs(genome1.aLinkGenes[i].Weight() * linkEnabledInGenome1 - 
+                           genome2.aLinkGenes[idx].Weight() * linkEnabledInGenome2);
+        ++coincident;
+      }
+    }
+
+    deltaW = deltaW / coincident;
+    return deltaW;
+  }
+
+  // Whether two genome belong to same species or not.
+  bool IsSameSpecies(Genome& genome1, Genome& genome2) {
+    double deltaD = Disjoint(genome1, genome2);
+    double deltaW = WeightDiff(genome1, genome2);
+    double delta = aCoeffDisjoint * deltaD + aCoeffWeightDiff * deltaW;
+
+    if (delta < aCompatThreshold) {
```

Comment:
Where does this ``aCompatThreshold = 1.0``  parameter come from? Depending on the calculation of the disjoint, excess, and weight difference this could be different.